src/controladores/produtos.js

```diff
@@ -0,0 +1,99 @@
+const knex = require("../conexao")
+
+const cadastrarProduto = async (req, res) => {
+    const { descricao, quantidade_estoque, valor, categoria_id } = req.body
+
+    try {
+        const categoriaIdValida = await knex('categorias').where({ id: categoria_id }).first()
+
+        if (!categoriaIdValida) {
+            return res.status(404).json({ mensagem: 'A categoria_id informada não foi encontrada.' })
+        }
+
+        const produto = await knex('produtos').insert({
+            descricao,
+            quantidade_estoque,
+            valor,
+            categoria_id
+        }).returning('*')
+
+        return res.status(201).json(produto[0])
+    } catch (error) {
+        return res.status(500).json({ mensagem: error.message })
+    }
+}
+
+const editarProduto = async (req, res) => {
+    const id = req.params.id;
+    const { descricao, quantidade_estoque, valor, categoria_id } = req.body;
+
+    try {
+        const produtoExistente = await knex("produtos").where("id", id).first();
+
+        if (!produtoExistente) {
+            return res.status(404).json({ mensagem: "Produto não encontrado." });
+        }
+
+        const produtoEditado = await knex("produtos")
+            .where("id", id)
+            .update({ descricao, quantidade_estoque, valor, categoria_id });
+
+        return res.status(204).send();
+    } catch (error) {
+        return res.status(500).json({ mensagem: error.message });
+    }
+};
+
+const listarProdutos = async (req, res) => {
+    const categoriaId = req.query.categoria_id;
+
+    let produtos;
+
+    if (categoriaId) {
+        const produtoExistente = await knex("produtos").where("categoria_id", categoriaId).first();
+
+        if (!produtoExistente) {
+            return res.status(404).json({ mensagem: "Categoria não encontrada." });
+        }
+
```

Comment:
Da forma como está sendo feita a validação do id da categoria, quando o usuário informa uma categoria que é válida, mas que não tem nenhum produto cadastrado está sendo retornada uma mensagem dizendo que a categoria não foi encontrada. Minha sugestão é trocar a validação pra ver se existe um produto (produtoExistente) por uma validação pra ver se existe uma categoria (categoriaExistente).

```suggestion
const categoriaExistente = await knex('categorias').where({ id: categoriaId }).first()

if (!categoriaExistente) {
     return res.status(404).json({ mensagem: 'O id de categoria informado não existe.' })
}
```